server/player/state/logged_in_state.py

```diff
@@ -0,0 +1,236 @@
+        serverid2 = request.findbytype(m02c4).value
+        game_server = self.player.server.find_server_by_id2(serverid2)
+        self.player.send(a00b0().setlength(10))
+        self.player.send(a0035().setserverdata(game_server))
+
+        modify_gameserver_whitelist('add', self.player, self.player.game_server)
+        self.player.game_server = game_server
+
+        # todo: add joined server state and enter it
+
+    @handles(packet=a00b3)
+    def handle_server_disconnect(self, request):  # server disconnect
+        # TODO: check on the real server if there's a response to this msg
+        # serverid2 = request.findbytype(m02c4).value
+        modify_gameserver_whitelist('remove', self.player, self.player.game_server)
+        self.player.game_server = None
+
+    @handles(packet=a0070)
+    def handle_chat(self, request):
+        message_type = request.findbytype(m009e).value
+
+        if message_type == 3:  # team
+            reply = a0070()
+            reply.findbytype(m009e).set(3)
+            reply.findbytype(m02e6).set('Unfortunately team messages are not yet supported. Use VGS for now.')
+            reply.findbytype(m02fe).set('taserver')
+            self.player.send(reply)
+
+        elif message_type == 6:  # private
+            addressed_player_name = request.findbytype(m034a).value
+            addressed_player = self.player.server.find_player_by(display_name=addressed_player_name)
+            if addressed_player:
+                request.content.append(m02fe().set(self.player.display_name))
+                request.content.append(m06de().set(self.player.tag))
+
+                self.player.send(request)
+
+                if self.player.id != addressed_player.id:
+                    addressed_player.send(request)
+
+        else:  # public
+            request.content.append(m02fe().set(self.player.display_name))
+            request.content.append(m06de().set(self.player.tag))
+
+            if self.player.game_server:
+                self.player.server.send_all_on_server(request, self.player.game_server)
+
+    @handles(packet=a0175)
+    def handle_promotion_code_redemption(self, request):
+        authcode = request.findbytype(m0669).value
+        if (self.player.login_name in self.player.server.accounts and
+                self.player.server.accounts[self.player.login_name].authcode == authcode):
+
+            self.player.server.accounts[self.player.login_name].password_hash = self.player.password_hash
+            self.player.server.accounts[self.player.login_name].authcode = None
+            self.player.server.accounts.save()
+            self.player.authenticated = True
+        else:
+            invalid_code_msg = a0175()
+            invalid_code_msg.findbytype(m02fc).set(0x00019646)  # message type
+            invalid_code_msg.findbytype(m0669).set(authcode)
+            self.player.send(invalid_code_msg)
+
+    @handles(packet=a018c)
+    def handle_votekick(self, request):
```

Comment:
python is meant to be self documenting. my biggest issue is that I never know what packet corresponds to what so this creates a clear mapping to the developer. and also make for much clearer logs and exceptions

Reply:
Ok, but then simplify the handler and handle_request by doing the following:

- storing the packet type in a "handles_packet" attribute and make that the only attribute in the decorator
- check only on that attribute when constructing the list of members in handle_request by using getattr(func, 'handles_packet', None) == type(request)
- raise an exception if there is more than one match and call the first function in the list
- put the decorator in the PlayerState file, because everyone who is going to import PlayerState is also going to be importing the decorator

Reply:
call the first function in the list? is that a good idea? i think i rather register the handlers inside the decorator and throw a runtime exception if a function tries to handle an already handled packet